src/riak_kv_qry_buffers.erl

```diff
@@ -660,6 +743,57 @@ tstamp() ->
 
 
 %% data ops
+
+-spec enkey([data_row()],
+            KeyFieldPositions::[non_neg_integer()],
+            OrdByFieldQualifiers::[{asc|desc, nulls_first|nulls_last}],
+            ChunkId::non_neg_integer()) ->
+                   [{{KeyOrd::riak_pb_ts_codec:ldbvalue(),
+                      ChunkId::non_neg_integer(),
+                      Idx::non_neg_integer()},
+                     data_row()}].
+enkey(Rows, KeyFieldPositions, OrdByFieldQualifiers, ChunkId)
+  when is_list(KeyFieldPositions) ->
+    %% 0. The new key is composed from fields appearing in the ORDER
+    %%    BY clause, and may therefore not work out to be unique. We
+    %%    now index the rows in the chunk to preserve the original
+    %%    order (because imposing any other order is even worse)
+    RowsIndexed = lists:zip(Rows, lists:seq(1, length(Rows))),
+    [begin
+         %% a. Form a new key from ORDER BY fields
+         KeyRaw = [lists:nth(Pos, Row) || Pos <- KeyFieldPositions],
+         %% b. Negate values in columns marked as DESC in ORDER BY clause
+         KeyOrd = [make_sorted_keys(F, AscDesc, Nulls)
+                   || {F, {AscDesc, Nulls}} <- lists:zip(KeyRaw, OrdByFieldQualifiers)],
+         %% c. Combine with chunk id and row idx to ensure uniqueness, and encode.
+         %%    Make sure the key proper goes first: ChunkId is a thing
+         %%    internal to query buffer, is incrementally increasing,
+         %%    whereas the chunk it refers to may certainly not be
+         %%    increasing because of out-of-order arrivals from the
+         %%    vnodes.
+         Key = {KeyOrd, ChunkId, Idx},
+         %% d. Encode the record (don't bother constructing a
+         %%    riak_object with metadata, vclocks):
+         {Key, Row}
+     end || {Row, Idx} <- RowsIndexed].
+
+make_sorted_keys([], desc, nulls_first) ->
```

Comment:
Simplified and *corrected* in 0544675.

Now nulls are sorted as described in https://drive.google.com/file/d/0B-2H_9GJ40JpMDAwY0VyWjkwSWc (page 617 in the pdf version of it). Specifically:

ASC NULLS LAST yields
```1, 2, 3, null, null```
ASC NULLS FIRST yields
```null, null, 1, 2, 3```
DESC NULLS LAST yields
```null, null, 3, 2, 1```
DESC NULLS FIRST yields
```3, 2, 1, null, null```

The test also needed to be updated: done in https://github.com/basho/riak_test/pull/1257/commits/c595fe7e0f5a8d8d53c18f34383bc499f5bae73b.